nixos/modules/services/backup/borgbackup.nix

```diff
@@ -0,0 +1,488 @@
+        ProtectSystem = "full";
+        #FIXME:
+        #ProtectSystem = "strict";
+        #ReadWritePaths =
+        #  [ "${userHome}/.config/borg" "${userHome}/.cache/borg" ]
+        #  # Borg needs write access to repo if it is not remote
+        #  ++ optional (isPath cfg.repo) cfg.repo;
+      };
+      environment = {
+        BORG_REPO = cfg.repo;
+      } // (mkPassEnv cfg) // cfg.environment;
+      inherit (cfg) startAt;
+    };
+
+  # The name arg is needed for mapAttrsToList
+  mkPassAssertion = name: cfg: {
+    assertion = with cfg.encryption;
+      mode != "none" -> passCommand != null || passphrase != null;
+    message =
+      ''passCommand or passphrase have to be specified ''
+      + ''because borgbackup.jobs.${name}.encryption != "none"'';
+  };
+
+  mkRepoService = name: cfg:
+    nameValuePair "borgbackup-repo-${name}" {
+      description = "Create BorgBackup repository ${name} directory";
+      script = ''
+        mkdir -p ${escapeShellArg cfg.path}
+        chown ${cfg.user}:${cfg.group} ${escapeShellArg cfg.path}
+      '';
+      serviceConfig = {
+        # The service's only task is to ensure that the specified path exists
+        Type = "oneshot";
+      };
+      wantedBy = [ "multi-user.target" ];
+      after = [ "multi-user.service" ];
+    };
+
+  mkAuthorizedKey = cfg: appendOnly: key:
+    let
+      restrictedArg = "--restrict-to-repository ${escapeShellArg cfg.path}";
+      appendOnlyArg = optionalString appendOnly "--append-only";
+      quotaArg = optionalString (cfg.quota != null) "--storage-quota ${cfg.quota}";
+      command = "borg serve ${restrictedArg} ${appendOnlyArg} ${quotaArg}";
+    in
+      ''command="${command}",restrict ${key}'';
+
+  mkUsersConfig = name: cfg: {
+    users.${cfg.user} = {
+      openssh.authorizedKeys.keyFiles = singleton
+        (pkgs.writeText "borgbackup.${cfg.user}.authorized_keys" (concatStringsSep "\n"
+          (map (mkAuthorizedKey cfg false) cfg.publicKeys
+          ++ map (mkAuthorizedKey cfg true) cfg.publicKeysAppendOnly)
+        ));
+      useDefaultShell = true;
+    };
+    groups.${cfg.group} = { };
+  };
+
+  # The name arg is needed for mapAttrsToList
+  mkKeysAssertion = name: cfg: {
+    assertion = cfg.publicKeys != [ ] || cfg.publicKeysAppendOnly != [ ];
+    message =
+      "borgbackup.repos.${name} does not make sense"
+      + " without at least one public key";
+  };
+
+in {
+  meta.maintainers = with maintainers; [ dotlambda ];
+
+  ###### interface
+
+  options.services.borgbackup.jobs = mkOption {
+    description = "Deduplicating backups using BorgBackup.";
+    default = { };
+    type = types.attrsOf (types.submodule (
+      { name, config, ... }: {
+        options = {
+
+          paths = mkOption {
+            type = with types; either path (nonEmptyListOf path);
+            description = "Paths to back up.";
+            example = "/home/user";
+            apply = x: if isList x then x else [ x ];
```

Comment:
I would prefer to have type just be "nonEmptyListOf path". Reasoning: the option and its documentation is in plural, so not using a list feels weird to me.

Reply:
It's just for convenience. I usually back up just one directory and this way I don't have to write the brackets. I can write `Path(s) to back up.` if you want :)

Reply:
It's fine. I'm just talking out loud :-)

Reply:
*thinking* out loud